src/dotnet-scaffolding/Microsoft.DotNet.Scaffolding.Helpers/Services/MsBuildInitializer.cs

```diff
@@ -1,19 +1,81 @@
+using System.Reflection;
+using System.Runtime.InteropServices;
+using System.Runtime.Loader;
 using Microsoft.Build.Locator;
 
 namespace Microsoft.DotNet.Scaffolding.Helpers.Services;
 
-internal static class MsBuildInitializer
+internal class MsBuildInitializer
 {
+    private readonly ILogger _logger;
+    public MsBuildInitializer(ILogger logger)
+    {
+        _logger = logger;
+    }
+
+    public void Initialize()
+    {
+        RegisterMsbuild();
+    }
 
     /// <summary>
-    ///use MsBuildLocator.RegisterDefaults() to register the newest VS .NET SDK and MSBuild.
+    /// find the newest dotnet sdk on disk first, if none found, use the MsBuildLocator.RegisterDefaults().
     /// </summary>
     /// <returns></returns>
-    internal static void RegisterMsbuild()
+    private void RegisterMsbuild()
     {
         if (!MSBuildLocator.IsRegistered)
         {
-            MSBuildLocator.RegisterDefaults();
+            // Path to the directory containing the SDKs
+            string sdkBasePath = GetDefaultSdkPath();
+            //register newest MSBuild from the newest dotnet sdk installed.
+            var sdkPath = Directory.GetDirectories(sdkBasePath)
+                                  .OrderByDescending(d => new DirectoryInfo(d).Name)
+                                  .FirstOrDefault();
+
+            if (!Directory.Exists(sdkBasePath) || string.IsNullOrEmpty(sdkPath))
```

Comment:
You check for the existence of sdkBasePath after you called Directory.GetDirectories() on it. Should this be called first? I think Directories.GetDirectories() will throw if the path doesn't exist

Reply:
yep not sure why I tried to combine these checks, separated them.